src/query/api/v1/httpd/handler.go

```diff
@@ -465,37 +464,62 @@ func (h *Handler) RegisterRoutes() error {
 
 			if !prevRoute {
 				if err := h.registry.Register(queryhttp.RegisterOptions{
-					Path:       custom.Route(),
-					Handler:    handler,
-					Methods:    methods(method),
-					Middleware: custom.Middleware(),
+					Path:               custom.Route(),
+					Handler:            handler,
+					Methods:            methods(method),
+					MiddlewareOverride: custom.MiddlewareOverride(),
 				}); err != nil {
 					return err
 				}
 			} else {
-				// TODO(rhall): we should figure out a way to allow merging middleware between the previous and custom
-				// handlers.
-				if custom.Middleware() != nil {
-					return fmt.Errorf("cannot apply middleware to an existing route: %s", custom.Route())
-				}
+				customMiddle[entry] = custom.MiddlewareOverride()
 				entry.Handler(handler)
 			}
 		}
 	}
 
+	// NB: the double http_handler was accidentally introduced and now we are
+	// stuck with it for backwards compatibility.
+	middleIOpts := instrumentOpts.SetMetricsScope(
+		h.options.InstrumentOpts().MetricsScope().SubScope("http_handler_http_handler"))
+
 	// Apply middleware after the custom handlers have overridden the previous handlers so the middleware functions
 	// are dispatched before the custom handler.
 	// req -> middleware fns -> custom handler -> previous handler.
-	for _, e := range h.registry.Entries() {
-		h := e.Route.GetHandler()
+	err = h.registry.Walk(func(route *mux.Route, router *mux.Router, ancestors []*mux.Route) error {
+		handler := route.GetHandler()
+		opts := middleware.Options{
+			InstrumentOpts: middleIOpts,
+			Route:          route,
+			Clock:          clockwork.NewRealClock(),
+			Logging: middleware.LoggingOptions{
+				Threshold: h.middlewareConfig.Logging.Threshold,
+			},
+			Metrics: middleware.MetricsOptions{
+				Config: h.middlewareConfig.Metrics,
+			},
+		}
+		override := h.registry.MiddlewareOpts(route)
+		if override != nil {
+			opts = override(opts)
+		}
+		if customMiddle[route] != nil {
```

Comment:
now we support a custom handler overriding the middleware options of a prev handler.